lib/ace/ext/papeeria/spellchecker.coffee

```diff
@@ -1,39 +1,217 @@
 
-    exports.SpellChecker = SpellChecker
+
+  # Sets up spellchecker popup and implements some routines
+  # to work on current in the editor.
+  setupSpellCheckerPopup = (editor) ->
+    PopupManager.init(editor)
+
+    # Check if current word is in corrections list.
+    # Call PopupManager.show if it's present, do nothing otherwise.
+    newPopup = ->
+      word = extractWord(editor)
+      spellChecker = new SpellChecker()
+      correctionsItem = spellChecker.getCorrections(word)
+      if correctionsItem
+        PopupManager.show(convertCorrectionList(correctionsItem))
+      return
+
+    # Bind newPopup to Alt-Enter editor shortcut.
+    editor.commands.addCommand({
+      name: "newPopup",
+      bindKey: "Alt-Enter",
+      exec: newPopup
+    })
     return
-)
\ No newline at end of file
+
+
+  # Get the word under the cursor.
+  # @param {Editor} editor: editor object.
+  # @return {String}: the current word.
+  extractWord = (editor) ->
+    session = editor.getSession()
+    wordRange = getCurrentWordRange(editor)
+    return session.getTextRange(wordRange)
+
+  # Returns Range object that describes the current word position.
+  # @param {Editor} editor: editor object.
+  # @return {Range}: current word range.
+  getCurrentWordRange = (editor) ->
+    session = editor.getSession()
+    row = editor.getCursorPosition().row
+    col = editor.getCursorPosition().column
+    return session.getWordRange(row, col)
+
+  # Convert an array of string to popup-eligible structure.
+  # @param {Array} corrections: array of strings with substitution options.
+  # @return {Array}: array of JSONs, actually.
+  convertCorrectionList = (corrections) ->
+    return ({caption: item, value: item} for item in corrections)
+
+
+  # Object contains init, show and detach functions for popup
+  # and some routines for inner use.
+  PopupManager =
+    # Bindings for editor keys with PopupManager functions.
+    commands:
+      "Esc": ->
+        PopupManager.detach()
+      "Up": ->
+        PopupManager.goTo("up")
+      "Down": ->
+        PopupManager.goTo("down")
+      "Return": ->
+        PopupManager.insertCorrection()
+
+    # Initial setup for popup.
+    # @param {Editor} editor: editor object.
+    init: (editor) ->
+      @editor = editor
+      @activated = false
+
+      HashHandler = require("ace/keyboard/hash_handler").HashHandler
+      AcePopup = require("ace/autocomplete/popup").AcePopup
+
+      @popup = new AcePopup(document.body)
+      @popup.setTheme(editor.getTheme())
+      @popup.setFontSize(editor.getFontSize())
+      @popup.on("click", (e) =>
+        @insertCorrection()
+        e.stop()
+      )
+
+      @session = editor.getSession()
+
+      @keyboard = new HashHandler()
+      @keyboard.bindKeys(@commands)
+
+      return
+
+    # Show popup in editor.
+    # @param {Array} options: list of corrections for the current word.
+    show: (options) ->
+      PopupManager.detach()
```

Comment:
If you maintain invariants "popup hidden == @activated is false", "popup show == @activated is true", "@activated is False == detach has been called or show has never been called" then you don't need to call `detach` here. It may not be always possible to maintain these invariants, e.g. when popup can hide so that you're not notified about that, however, if you're not aware of such "backdoors" then I recommend to try maintaining the invariant.